### ISSUE TITLE
plugin/hook for Pyodide?

### ISSUE THREAD

**Issue by ryanking13** [pyodide/pyodide#3316]
Recently, we are getting some requests and ideas of accessing and modifying Pyodide internal objects before initialization:

- Accessing and modifying filesystem: #3274
- Caching compiled .py sources on runtime: https://github.com/pyodide/pyodide/issues/3269#issuecomment-1335252076

So I think it would be nice if we can provide some plugin/hook (I'm not sure this is a correct term for this) interface that users can easily add custom operations before initialization, without modifying or monkey patching Pyodide. Ideally, users can write useful plugins/hooks themselves, and maybe if we later found some of those hooks are critical for most use cases, we can think of porting them into Pyodide.

Below is my very rough idea of the plugin/hook interface. I haven't thought through this, so any ideas are welcome.

1. Add plugins parameter to loadPyodide

```typescript
type PyodidePlugin= {
  preInit?: ...
  preRun?: ...
  onFSInit?: ...
  onPyodideInit?: ...
  ...
}

function loadPyodide(options: {
   ...
   plugins?: PyodidePlugin[]
   ...
}
```

2. Run each hook inside plugins before/after specific steps during initialization

```typescript
function loadPyodide(...) {
   // after initializing module object
   plugins.forEach(hook=> { if (hook.hasOwnProperty("onModuleInit")) hook.onModuleInit(module) })
   ...
   // after initializing Pyodide
   plugins.forEach(hook=> { if (hook.hasOwnProperty("onPyodideInit")) hook.onPyodideInit() })
}
```

Maybe this is also partially related to how we can implement #3260.

cc: @joemarshall @madhur-tandon  @antocuni

**Comment by joemarshall** [comment]
I'd like hook on package loading also, I.e. onloadpackage

**Comment by joemarshall** [comment]
And onfsinit/onpyodideinit should be clearly named so it is obvious whether they get called before or after. E.g. onPyodideInited and onFSAvailable

**Comment by ryanking13** [comment]
> I'd like hook on package loading also, I.e. onloadpackage

Yes, that sounds good to me. Probably we need to discuss whether we should implement it in micropip or in pyodide.loadPackage.

> And onfsinit/onpyodideinit should be clearly named so it is obvious whether they get called before or after. E.g. onPyodideInited and onFSAvailable

Yeah, the name of hooks that I wrote is just examples.
